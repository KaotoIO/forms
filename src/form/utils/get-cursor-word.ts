const SPACE = ' ';

/**
 * Utility function to get the word at the cursor position in a text input.
 */
export const getCursorWord = (
  inputValue: string | number,
  cursorPosition: number | null = 0,
): { word: string; cursorStart: number; cursorEnd: number } => {
  const value = typeof inputValue === 'string' ? inputValue : String(inputValue);
  const cursor = cursorPosition ?? 0;

  // Find the start of the word (scan left from cursor)
  let wordStart = Math.max(cursor, 0);
  while (wordStart > 0 && value[wordStart - 1] !== SPACE) {
    wordStart--;
  }

  // Find the end of the word (scan right from cursor)
  let wordEnd = Math.min(cursor, value.length);
  while (wordEnd < value.length && value[wordEnd] !== SPACE) {
    wordEnd++;
  }

  const word = value.slice(wordStart, wordEnd).trim();

  return { word, cursorStart: wordStart, cursorEnd: wordEnd };
};
