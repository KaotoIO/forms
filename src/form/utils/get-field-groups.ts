import { JSONSchema4 } from 'json-schema';
import { extractGroup } from './get-tagged-field-from-string';
import { isDefined } from './is-defined';

interface FieldGroups {
  common: Record<string, JSONSchema4>;
  groups: [string, Record<string, JSONSchema4>][];
}

export const getFieldGroups = (properties?: JSONSchema4['properties']): FieldGroups => {
  if (!isDefined(properties)) return { common: {}, groups: [] };

  const groupedProperties = Object.entries(properties).reduce(
    (acc, [name, definition]) => {
      // "$comment": "group:advanced" or "$comment": "group:consumer (advanced)"
      const group = extractGroup('group', definition.$comment);

      if (group === '' || group === 'common' || group === 'producer' || group === 'consumer') {
        acc.common[name] = definition;
      } else {
        acc.groups[group] ??= {};
        acc.groups[group][name] = definition;
      }

      return acc;
    },
    { common: {}, groups: {} } as {
      common: Record<string, JSONSchema4>;
      groups: Record<string, Record<string, JSONSchema4>>;
    },
  );

  /** Prioritize advanced properties */
  const groupArray = Object.entries(groupedProperties.groups).sort((a, b) => {
    if (a[0] === 'advanced') {
      return 1;
    } else if (b[0] === 'advanced') {
      return -1;
    }
    return 0;
  });

  return { common: groupedProperties.common, groups: groupArray };
};
