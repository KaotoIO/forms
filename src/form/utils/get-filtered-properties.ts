import { JSONSchema7 } from 'json-schema';
import { isDefined } from './is-defined';

/**
 * Extracts the schema recursively containing only the filtered properties.
 */
export function getFilteredProperties(
  properties: JSONSchema7['properties'],
  filter: string,
  omitFields?: string[],
): JSONSchema7['properties'] {
  if (!isDefined(properties)) return {};

  const filteredFormSchema = Object.entries(properties).reduce(
    (acc, [property, definition]) => {
      if (!omitFields?.includes(property)) {
        if (typeof definition === 'object' && definition['type'] === 'object' && isDefined(definition['properties'])) {
          const subFilteredSchema = getFilteredProperties(definition['properties'], filter);
          if (subFilteredSchema && Object.keys(subFilteredSchema).length > 0) {
            acc![property] = { ...definition, properties: subFilteredSchema };
          }
        } else if (property.toLowerCase().includes(filter)) {
          acc![property] = definition;
        }
      }

      return acc;
    },
    {} as JSONSchema7['properties'],
  );

  return filteredFormSchema;
}
