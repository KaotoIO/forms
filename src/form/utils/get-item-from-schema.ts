import { JSONSchema4 } from 'json-schema';
import { resolveSchemaWithRef } from './resolve-schema-with-ref';

export const getItemFromSchema = (schema: JSONSchema4, definitions: Record<string, JSONSchema4>) => {
  const resolvedSchema = resolveSchemaWithRef(schema, definitions);
  const defaultValue = resolvedSchema.default;
  const properties = resolvedSchema.properties ?? {};
  const required = Array.isArray(resolvedSchema.required) ? resolvedSchema.required : [];

  switch (resolvedSchema.type) {
    case 'string':
      return defaultValue ?? '';
    case 'boolean':
      return defaultValue ?? false;
    case 'number':
      return defaultValue ?? 0;
    case 'object':
      return Object.entries(properties).reduce(
        (acc, [key, value]) => {
          if (required.includes(key)) {
            acc[key] = getItemFromSchema(value, definitions);
          }

          return acc;
        },
        {} as Record<string, unknown>,
      );
    default:
      return {};
  }
};
