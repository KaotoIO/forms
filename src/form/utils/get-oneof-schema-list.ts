import { JSONSchema7, JSONSchema7Definition } from 'json-schema';
import { camelCaseToSpaces } from './camel-case-to-space';
import { isDefined } from './is-defined';
import { resolveSchemaWithRef } from './resolve-schema-with-ref';

export interface OneOfSchemas {
  name: string;
  description?: string;
  schema: JSONSchema7;
}

export const getOneOfSchemaList = (
  oneOfList: JSONSchema7[],
  definitions: JSONSchema7Definition = {},
): OneOfSchemas[] => {
  const list = oneOfList
    /** Ignore the `not` schemas */
    .filter((oneOfSchema) => !isDefined(oneOfSchema.not))
    .map((schema, index) => {
      const resolvedSchema = resolveSchemaWithRef(schema, definitions);

      let name = resolvedSchema.title;
      let description = resolvedSchema.description;

      const schemaProps = Object.keys(resolvedSchema.properties ?? {});
      const isSinglePropertySchema = schemaProps.length === 1;
      if (isSinglePropertySchema) {
        const singlePropertySchema = resolveSchemaWithRef(
          resolvedSchema.properties![schemaProps[0]] as JSONSchema7,
          definitions,
        );
        name ??= singlePropertySchema.title ?? camelCaseToSpaces(schemaProps[0], { capitalize: true });
        description ??= singlePropertySchema.description;
      }

      name ??= `Schema ${index}`;

      return { name, description, schema: resolvedSchema };
    });

  /** Handling cases where an EIP, Language or Dataformat could be represented in 2 different ways  */
  if (list.length === 2) {
    const [firstSchema, secondSchema] = list;
    if (firstSchema.schema.type === 'string' && secondSchema.schema.type === 'object') {
      firstSchema.name = 'Simple';
      secondSchema.name = 'Advanced';
    }
  }

  return list;
};
