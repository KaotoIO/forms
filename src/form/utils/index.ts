export * from './camel-case-to-space';
export * from './camel-random-id';
export * from './get-field-groups';
export * from './get-filtered-properties';
export * from './get-item-from-schema';
export * from './get-value';
export * from './is-defined';
export * from './is-raw-string';
export * from './resolve-schema-with-ref';
export * from './set-value';
