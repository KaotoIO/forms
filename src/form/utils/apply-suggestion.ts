import { Suggestion } from '../models/suggestions';
import { getCursorWord } from './get-cursor-word';

export const applySuggestion = (
  suggestion: Suggestion,
  inputValue: string | number,
  cursorPosition?: number | null,
): { newValue: string; cursorPosition: number } => {
  const cursor = cursorPosition ?? 0;

  const { word, cursorStart, cursorEnd } = getCursorWord(inputValue, cursorPosition);

  const firstSection = inputValue.toString().substring(0, cursorStart);
  const lastSection = inputValue.toString().substring(cursorEnd);
  const newValue = firstSection + suggestion.value + lastSection;

  return {
    newValue,
    cursorPosition: cursorStart + suggestion.value.length,
  };
};
