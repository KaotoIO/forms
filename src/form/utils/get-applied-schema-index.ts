import { JSONSchema7, JSONSchema7Definition } from 'json-schema';
import { OneOfSchemas } from './get-oneof-schema-list';
import { weightSchemaAgainstModel } from './weight-schemas-against-model';

export const getAppliedSchemaIndex = (
  model: unknown,
  oneOfSchemaList: OneOfSchemas[],
  definitions: JSONSchema7Definition = {},
): number => {
  const schemaPoints = oneOfSchemaList
    .map(({ schema }, index) => {
      const points = weightSchemaAgainstModel(model, schema, definitions);
      return { index, points };
    })
    .filter((weigthedSchema) => weigthedSchema.points > 0)
    .sort((a, b) => b.points - a.points);

  return schemaPoints.length > 0 ? schemaPoints[0].index : -1;
};
