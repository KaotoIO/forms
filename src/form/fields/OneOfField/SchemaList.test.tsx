import { render, screen, fireEvent } from '@testing-library/react';
import { SchemaList } from './SchemaList';
import { OneOfSchemas } from '../../utils/get-oneof-schema-list';

describe('SchemaList', () => {
  const mockSchemas: OneOfSchemas[] = [
    {
      name: 'Schema 1',
      description: 'First schema',
      schema: { type: 'object', properties: { name: { type: 'string' } } },
    },
    {
      name: 'Schema 2',
      description: 'Second schema',
      schema: { type: 'object', properties: { value: { type: 'number' } } },
    },
  ];

  const manySchemas = [...mockSchemas];
  manySchemas.push(
    { name: 'Schema 3', description: 'Third schema', schema: { type: 'string' } },
    { name: 'Schema 4', description: 'Fourth schema', schema: { type: 'string' } },
    { name: 'Schema 5', description: 'Fifth schema', schema: { type: 'string' } },
    { name: 'Schema 6', description: 'Sixth schema', schema: { type: 'string' } },
  );

  const mockOnChange = jest.fn();
  const mockOnCleanInput = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with basic props when schema is selected', () => {
    const { container } = render(
      <SchemaList
        propName="testProp"
        selectedSchema={mockSchemas[0]}
        schemas={mockSchemas}
        onChange={mockOnChange}
        onCleanInput={mockOnCleanInput}
        placeholder="Select schema"
        aria-label="Test schema list"
        data-testid="test-schema-list"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly with basic props when no schema is selected', () => {
    const { container } = render(
      <SchemaList
        propName="testProp"
        selectedSchema={undefined}
        schemas={mockSchemas}
        onChange={mockOnChange}
        onCleanInput={mockOnCleanInput}
        placeholder="Select schema"
        aria-label="Test schema list"
        data-testid="test-schema-list"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should use Typeahead when items > 5, none selected', () => {
    const { container } = render(
      <SchemaList
        propName="testProp"
        selectedSchema={undefined}
        schemas={manySchemas}
        onChange={mockOnChange}
        data-testid="test-schema-list"
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should use Typeahead when items > 5, one selected', () => {
    const { container } = render(
      <SchemaList
        propName="testProp"
        selectedSchema={manySchemas[4]}
        schemas={manySchemas}
        onChange={mockOnChange}
        onCleanInput={mockOnCleanInput}
        data-testid="test-schema-list"
      />,
    );

    expect(container).toMatchSnapshot();

    // clear the selection
    const clearButton = screen.getByTestId('test-schema-list__clear');
    fireEvent.click(clearButton);

    expect(mockOnCleanInput).toHaveBeenCalled();
  });

  it('should handle schema selection and change', () => {
    render(
      <SchemaList
        propName="testProp"
        selectedSchema={manySchemas[0]}
        schemas={manySchemas}
        onChange={mockOnChange}
        aria-label="Test schema list"
        data-testid="test-schema-list"
      />,
    );

    const changeButton = screen.getByRole('button', { name: 'Test schema list toggle' });
    fireEvent.click(changeButton);

    fireEvent.click(screen.getByRole('option', { name: 'option schema 2' }));

    expect(mockOnChange).toHaveBeenCalledWith({
      name: 'Schema 2',
      description: 'Second schema',
      schema: { type: 'object', properties: { value: { type: 'number' } } },
    });
  });
});

