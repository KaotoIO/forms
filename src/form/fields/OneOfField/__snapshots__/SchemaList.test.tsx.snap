// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaList should render correctly with basic props when no schema is selected 1`] = `
<div>
  <div
    aria-labelledby="testProp-legend"
    class="pf-v6-c-form__group"
    role="group"
  >
    <div
      class="pf-v6-c-form__group-control pf-m-stack"
    >
      <div
        aria-label="Test schema list"
        class="pf-v6-c-toggle-group"
        data-testid="test-schema-list"
        id="testProp"
        role="group"
      >
        <div
          class="pf-v6-c-toggle-group__item"
        >
          <button
            aria-pressed="false"
            class="pf-v6-c-toggle-group__button"
            id="Schema 1"
            type="button"
          >
            <span
              class="pf-v6-c-toggle-group__text"
            >
              Schema 1
            </span>
          </button>
        </div>
        <div
          class="pf-v6-c-toggle-group__item"
        >
          <button
            aria-pressed="false"
            class="pf-v6-c-toggle-group__button"
            id="Schema 2"
            type="button"
          >
            <span
              class="pf-v6-c-toggle-group__text"
            >
              Schema 2
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SchemaList should render correctly with basic props when schema is selected 1`] = `
<div>
  <div
    aria-labelledby="testProp-legend"
    class="pf-v6-c-form__group"
    role="group"
  >
    <div
      class="pf-v6-c-form__group-control pf-m-stack"
    >
      <div
        aria-label="Test schema list"
        class="pf-v6-c-toggle-group"
        data-testid="test-schema-list"
        id="testProp"
        role="group"
      >
        <div
          class="pf-v6-c-toggle-group__item"
        >
          <button
            aria-pressed="true"
            class="pf-v6-c-toggle-group__button pf-m-selected"
            id="Schema 1"
            type="button"
          >
            <span
              class="pf-v6-c-toggle-group__text"
            >
              Schema 1
            </span>
          </button>
        </div>
        <div
          class="pf-v6-c-toggle-group__item"
        >
          <button
            aria-pressed="false"
            class="pf-v6-c-toggle-group__button"
            id="Schema 2"
            type="button"
          >
            <span
              class="pf-v6-c-toggle-group__text"
            >
              Schema 2
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SchemaList should use Typeahead when items > 5, none selected 1`] = `
<div>
  <div
    aria-labelledby="testProp-legend"
    class="pf-v6-c-form__group"
    role="group"
  >
    <div
      class="pf-v6-c-form__group-control pf-m-stack"
    >
      <div
        class="pf-v6-c-menu-toggle pf-m-full-width pf-m-typeahead"
        id="testProp"
      >
        <div
          class="pf-v6-c-text-input-group pf-m-plain"
        >
          <div
            autocomplete="off"
            class="pf-v6-c-text-input-group__main"
            data-testid="test-schema-list-typeahead-select-input"
            id="testProp-typeahead-select-input"
          >
            <span
              class="pf-v6-c-text-input-group__text"
            >
              <input
                aria-label="Type to filter"
                class="pf-v6-c-text-input-group__text-input"
                placeholder="Select or write an option"
                type="text"
                value=""
              />
            </span>
          </div>
          <div
            class="pf-v6-c-text-input-group__utilities"
          />
        </div>
        <button
          aria-expanded="false"
          aria-label="undefined toggle"
          class="pf-v6-c-menu-toggle__button"
          data-ouia-component-id="OUIA-Generated-MenuToggle-typeahead-1"
          data-ouia-component-type="PF6/MenuToggle"
          data-ouia-safe="true"
          tabindex="-1"
          type="button"
        >
          <span
            class="pf-v6-c-menu-toggle__controls"
          >
            <span
              class="pf-v6-c-menu-toggle__toggle-icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v6-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                />
              </svg>
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`SchemaList should use Typeahead when items > 5, one selected 1`] = `
<div>
  <div
    aria-labelledby="testProp-legend"
    class="pf-v6-c-form__group"
    role="group"
  >
    <div
      class="pf-v6-c-form__group-control pf-m-stack"
    >
      <div
        class="pf-v6-c-menu-toggle pf-m-full-width pf-m-typeahead"
        id="testProp"
      >
        <div
          class="pf-v6-c-text-input-group pf-m-plain"
        >
          <div
            autocomplete="off"
            class="pf-v6-c-text-input-group__main"
            data-testid="test-schema-list-typeahead-select-input"
            id="testProp-typeahead-select-input"
          >
            <span
              class="pf-v6-c-text-input-group__text"
            >
              <input
                aria-label="Type to filter"
                class="pf-v6-c-text-input-group__text-input"
                placeholder="Select or write an option"
                type="text"
                value="Schema 5"
              />
            </span>
          </div>
          <div
            class="pf-v6-c-text-input-group__utilities"
          >
            <button
              aria-label="Clear input value"
              class="pf-v6-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF6/Button"
              data-ouia-safe="true"
              data-testid="test-schema-list__clear"
              type="button"
            >
              <span
                class="pf-v6-c-button__icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v6-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 352 512"
                  width="1em"
                >
                  <path
                    d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
        <button
          aria-expanded="false"
          aria-label="undefined toggle"
          class="pf-v6-c-menu-toggle__button"
          data-ouia-component-id="OUIA-Generated-MenuToggle-typeahead-2"
          data-ouia-component-type="PF6/MenuToggle"
          data-ouia-safe="true"
          tabindex="-1"
          type="button"
        >
          <span
            class="pf-v6-c-menu-toggle__controls"
          >
            <span
              class="pf-v6-c-menu-toggle__toggle-icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v6-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                />
              </svg>
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
